_posts/2019-03-19-til-validations-and-constraints.md

```diff
@@ -0,0 +1,229 @@
+phase will be looking at when we need to apply some constraints.
+
+### Constraints
+You might have gotten to this point and thought, why are there differences?
+Let's think about many apps that we use. When we sign up for an application, are
+we allowed to signup with the same email as another user? (Hint: the answer should
+always be no.) The reason is because developers want to be able to distinguish users
+between one another and ensure that they can send the right person the right information.
+
+So why couldn't we just have a validation for uniqueness? Well remember the definition
+we learned about validations, we perform the validation prior to checking the database.
+If we were to have a validation for uniqueness, that would mean everything is
+unique even if you're adding duplicates since it doesn't look at the database.
+
+A constraint is a rule that is enforced **by the database**. When an application,
+is finished checking through its validations, it then wants to ensure that it can be
+saved into the database just fine and isn't breaking any of the constraints.
+Let's add our first constraint, uniqueness!
+```elixir
+user
+|> cast(attrs, [:display_name, :email, :password])
+|> validate_required([:display_name, :email, :password])
+|> unique_constraint(:email)
+```
+
+If we try to add the user to our database, all looks good the first time:
+```elixir
+iex(2)> user = %User{}
+iex(3)> attrs = %{display_name: "prince", email: "prince@test.com", password: "super_secret"}
+iex(4)> user |> User.changeset(attrs) |> Repo.insert()
+{:ok,
+ %MyCoolWebApp.Accounts.User{
+   __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
+   display_name: "prince",
+   email: "prince@test.com",
+   id: 1,
+   inserted_at: ~N[2019-03-18 01:41:34],
+   password: "super_secret",
+   password_hash: "$argon2i$v=19$m=65536,t=6,p=1$bhjgmBs9/gYcM2L5Z5sL/g$Z+4D7NIaauU+jwhdYRY4hz0adUdhjAJK6CwYk1AOJdE",
+   updated_at: ~N[2019-03-18 01:41:34]
+ }}
+```
+
+We want to make sure no duplicates get saved, so let's try sending the same thing again:
+```elixir
+iex(2)> user = %User{}
+iex(3)> attrs = %{display_name: "prince", email: "prince@test.com", password: "super_secret"}
+iex(4)> user |> User.changeset(attrs) |> Repo.insert()
+{:ok,
+ %MyCoolWebApp.Accounts.User{
+   __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
+   display_name: "prince",
+   email: "prince@test.com",
+   id: 2,
+   inserted_at: ~N[2019-03-18 01:43:57],
+   password: "super_secret",
+   password_hash: "$argon2i$v=19$m=65536,t=6,p=1$H+Fq/IPW+M0YPHOZxMs13Q$ne+jDkwfcOigT8TKDIBYJjVwNdaNkzF/hc7YcRXRItY",
+   updated_at: ~N[2019-03-18 01:43:57]
+ }}
+```
+
+That's weird. It didn't show us an error? That's because Ecto doesn't know it is
+supposed to show an error. If you add a constraint to your `changeset/3`, you **must**
+put it into the migrations, the database level. Ecto only checks constraints if they are enforced
```

Comment:
When you say "put it into the migrations", what do you mean? If we add _any_ kind of constraint then we need to add an index on the db level? Or are there different ways to add your constraint to the db depending on type of constraint?